### PR TITLE
Saving Resource Variables without Copy

### PR DESCRIPTION
**Pull request by philipphack**
Provides a solution for the increase in memory usage when saving models with large sparsely accessed resource variables by avoiding the copying of the variables. The kernel of `ReadVariableOp` is modified to convert variables that were sparsely accessed from copy-on-read mode to copy-on-write mode under an exclusive lock. The conversion is disabled in the default setting and only occurs when the Python function `_read_variable_op` is called from class `ResourceVariableSaveable` during the saving of resource variables.

See issue [#54104](https://github.com/tensorflow/tensorflow/issues/54105).

### CODE REVIEW

**Comment by wangpengmit** [comment]
I don't like that `ReadVariableOp` does this COR-to-COW conversion as a side effect (even if it only happens when `no_copy` is true). How about adding a new op `TurnOffCopyOnRead` to do this job?

**Comment by philipphack** [comment]
Thanks for the comment. I've introduced a new `ReadVariableWithoutCopyOp`, which is exclusively called when saving resource variables. For copy-on-write variables, the op behaves like `ReadVariableOp`. For copy-on-read variables, the kernel again acquires an exclusive mutex, but avoids the side effect by directly reading the variable instead of changing its type.

File: tensorflow/core/kernels/resource_variable_ops.cc [hunk 1]
```diff
@@ -214,17 +219,61 @@ void ReadVariablesOp::Compute(OpKernelContext* ctx) {
...
+    // If the variable is in copy-on-read mode, obtain an exclusive
+    // lock on the variable before reading it without making a copy
+    mutex_lock ml(*variable->mu());
+    const Tensor* t = variable->tensor();
+    OP_REQUIRES(
+        ctx, dtype_ == t->dtype(),
+        errors::InvalidArgument(
+            "Trying to read variable with wrong dtype. Expected ",
+            DataTypeString(dtype_), " got ", DataTypeString(t->dtype())));
+    ctx->set_output(0, *t);
```
Comment (wangpengmit) [inline tensorflow/core/kernels/resource_variable_ops.cc:245]: The output tensor will be unstable. When the COR variable's buffer got modified, the output tensor's value will silently change, which breaks TF tensors' immutability guarantee.

My preferred solution is to add a `TurnOffCopyOnRead` (or `DisableCopyOnRead`) op to turn off a variable's copy_on_read_mode.

**Comment by philipphack** [comment]
The new `DisableCopyOnReadOp` is called when a resource variable is saved.

**Comment by wangpengmit** [comment]
Thanks for the update! Overall looks good, except for some minor issues.

**Comment by philipphack** [comment]
I can't reproduce the PyLint error, and I also don't see any lines that are longer than 80 characters in the diff.

**Comment by wangpengmit** [comment]
Maybe the test was against an older version. Or maybe it's caused by whitespaces. (CCing @mihaimaruseac who may know more.)

**Comment by mihaimaruseac** [comment]
Rerunning the pylint job, seems to have been using an old version?

Anyway, if still broken, please ignore.

File: tensorflow/core/kernels/resource_variable_ops.cc [hunk 2]
```diff
@@ -358,6 +359,31 @@ REGISTER_KERNEL_BUILDER(
...
+                  ", status error message=", status.error_message()));
+  // If the variable is currently in copy-on-read mode, its refcount is 1
+  if (variable->copy_on_read_mode.load()) {
+    // Obtain an exclusive lock on the variable and change the access mode
+    mutex_lock ml(*variable->mu());
+    variable->copy_on_read_mode.store(false);
+  }
+}
+
+REGISTER_KERNEL_BUILDER(Name("DisableCopyOnReadOp").Device(DEVICE_CPU),
```
Comment (wangpengmit) [inline tensorflow/core/kernels/resource_variable_ops.cc:381]: Op name needs to be updated.

File: tensorflow/python/ops/resource_variable_ops.py [hunk 3]
```diff
@@ -670,10 +670,15 @@ def _map_resources(self, save_options):
     resource_map = {self.handle: new_variable.handle}
     return obj_map, resource_map
 
-  def _read_variable_op(self):
+  def _read_variable_op(self, no_copy = False):
+    """Reads the value of the variable. If the variable is in copy-on-read mode
```
Comment (wangpengmit) [inline tensorflow/python/ops/resource_variable_ops.py:674]: The docstring needs to be in the format of the [Google Python Style Guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#383-functions-and-methods) to pass linters.

**Review by wangpengmit — APPROVED** [review]

File: tensorflow/python/ops/resource_variable_ops.py [hunk 4]
```diff
@@ -670,10 +670,22 @@ def _map_resources(self, save_options):
...
```
Comment (wangpengmit) [inline tensorflow/python/ops/resource_variable_ops.py:674]: There needs to be a one-line summary.

**Review by wangpengmit — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/resource_variable_ops.py [hunk 5]
```diff
@@ -710,6 +723,21 @@ def read_value(self):
     # specifies instead of the device where the variable is.
     return array_ops.identity(value)
 
+  def read_value_no_copy(self):
+    """Constructs an op which reads the value of this variable
```
Comment (wangpengmit) [inline tensorflow/python/ops/resource_variable_ops.py:727]: One-liner here also.

File: tensorflow/python/ops/resource_variable_ops.py [hunk 6]
```diff
@@ -710,6 +723,21 @@ def read_value(self):
...
+    without making a copy even when the variable has been previously
+    sparsely accessed. Variables that are in copy-on-read mode will
+    be converted to copy-on-write mode.
+
+    Returns:
+     the read operation.
```
Comment (wangpengmit) [inline tensorflow/python/ops/resource_variable_ops.py:733]: Should be "The value of the variable." (`read_value`'s docstring is outdated.)

**Review by wangpengmit — CHANGES_REQUESTED** [review]

**Review by wangpengmit — APPROVED** [review]

**Comment by wangpengmit** [comment]
There are some `api_compatibility_test` CI failures. Please run `bazel run //tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True` to update the API golden files.

**Comment by philipphack** [comment]
This fails with `ModuleNotFoundError: No module named 'tensorflow.core.kernels'`.

**Comment by wangpengmit** [comment]
Seems your Python paths are not set up properly. I'll try fixing the CI errors internally.

**Pull request merged**